### PR TITLE
improvement: Replace usage of Timex.shift with builtin Calendar functions

### PR DESCRIPTION
**Pull request by mbuhot**
Drops the dependency on `timex` and replaces usage of `Timex.shift` with `DateTime.add`, `Date.add` and some custom logic for working with months and years.

# Contributor checklist

- [ ] Chores
- [x] Documentation changes
- [ ] Features include unit/acceptance tests
- [x] Refactoring
- [x] Update dependencies

### CODE REVIEW

File: lib/ash/query/function/ago.ex [hunk 1]
```diff
@@ -3,28 +3,66 @@ defmodule Ash.Query.Function.Ago do
   Subtracts the given interval from the current time in UTC.
 
   For example:
-     deleted_at > ago(7, :days)
+     deleted_at > ago(7, :day)
```
Comment (mbuhot) [inline lib/ash/query/function/ago.ex:6]: I think this example was incorrect - `:days` isn't handled base the original `case` expressions.

File: lib/ash/query/function/ago.ex [hunk 2]
```diff
@@ -3,28 +3,66 @@ defmodule Ash.Query.Function.Ago do
...
-        :second -> :seconds
-        :millisecond -> :milliseconds
-        :microsecond -> :microseconds
+    now = DateTime.utc_now()
+    shifted = datetime_add(now, -factor, interval)
+    {:known, shifted}
+  end
+
+  @doc false
+  def datetime_add(datetime, amount_to_add, unit)
```
Comment (mbuhot) [inline lib/ash/query/function/ago.ex:22]: I added this as a public function to allow calling it from tests. Is it OK to keep it here? Or should it be in a helpers/utility module of some kind?

File: lib/ash/query/function/ago.ex [hunk 3]
```diff
@@ -3,28 +3,66 @@ defmodule Ash.Query.Function.Ago do
...
+
+  def datetime_add(datetime, amount_to_add, :week) do
+    datetime_add(datetime, 7 * amount_to_add, :day)
+  end
+
+  def datetime_add(datetime, amount_to_add, :month) do
+    months_since_zero = datetime.year * 12 + datetime.month - 1 + amount_to_add
+    year = div(months_since_zero, 12)
+    {year, month} =
+      case rem(months_since_zero, 12) do
```
Comment (mbuhot) [inline lib/ash/query/function/ago.ex:48]: There's some complexity here to handle dates earlier than year 0. Not sure if it can be simplified further eg by not allowing negative year dates? `Timex` would return `{:error, :shift_to_invalid_date}` for negative year dates.

File: lib/ash/query/function/ago.ex [hunk 4]
```diff
@@ -3,28 +3,66 @@ defmodule Ash.Query.Function.Ago do
...
+    # clamp to last day of month
+    last_day = Calendar.ISO.days_in_month(year, month)
+    day = min(datetime.day, last_day)
+    %{datetime | year: year, month: month, day: day}
+  end
+
+  def datetime_add(datetime = %{month: month, day: day}, amount_to_add, :year) do
+    result = %{datetime | year: datetime.year + amount_to_add}
+    # Adjust 29-Feb to 01-March when not a leap year
+    if month == 2 and day == 29 and not Calendar.ISO.leap_year?(result.year) do
```
Comment (mbuhot) [inline lib/ash/query/function/ago.ex:62]: Preserving the Timex behaviour here. Shifting 29th Feb by years to a non-leapyear adjusts forward to 1st of March.

File: mix.lock [hunk 5]
```diff
@@ -41,8 +39,6 @@
   "ssl_verify_fun": {:hex, :ssl_verify_fun, "1.1.6", "cf344f5692c82d2cd7554f5ec8fd961548d4fd09e7d22f5b62482e5aeaebd4b0", [:make, :mix, :rebar3], [], "hexpm", "bdb0d2471f453c88ff3908e7686f86f9be327d065cc1ec16fa4540197ea04680"},
   "stream_data": {:hex, :stream_data, "0.5.0", "b27641e58941685c75b353577dc602c9d2c12292dd84babf506c2033cd97893e", [:mix], [], "hexpm", "012bd2eec069ada4db3411f9115ccafa38540a3c78c4c0349f151fc761b9e271"},
   "telemetry": {:hex, :telemetry, "1.1.0", "a589817034a27eab11144ad24d5c0f9fab1f58173274b1e9bae7074af9cbee51", [:rebar3], [], "hexpm", "b727b2a1f75614774cff2d7565b64d0dfa5bd52ba517f16543e6fc7efcc0df48"},
-  "timex": {:hex, :timex, "3.7.5", "3eca56e23bfa4e0848f0b0a29a92fa20af251a975116c6d504966e8a90516dfd", [:mix], [{:combine, "~> 0.10", [hex: :combine, repo: "hexpm", optional: false]}, {:gettext, "~> 0.10", [hex: :gettext, repo: "hexpm", optional: false]}, {:tzdata, "~> 1.0", [hex: :tzdata, repo: "hexpm", optional: false]}], "hexpm", "a15608dca680f2ef663d71c95842c67f0af08a0f3b1d00e17bbd22872e2874e4"},
```
Comment (mbuhot) [inline mix.lock:44]: Unlocked timex and direct dependencies `gettext`, `combine` and `tzdata` which don't appear to be dependencies of any other packages.

File: test/query/function/ago_test.exs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  describe "ago query function" do
+    test "Years ago" do
+      today = Date.utc_today()
+      assert {:known, %DateTime{} = datetime} = Ago.evaluate(%{arguments: [1, :year]})
+      assert datetime.year == today.year - 1
+    end
+  end
+
+  describe "datetime_add helper" do
```
Comment (mbuhot) [inline test/query/function/ago_test.exs:14]: I've tried to focus the tests on the end-of-month and leap-year edge cases.

**Comment by jimsynz** [comment]
Mike you absolute legend.

**Comment by zachdaniel** [comment]
🔥 this looks awesome! Love to see those dependencies go away!

**Comment by zachdaniel** [comment]
looks like some credo/formatting to fix and then its good to go :)

File: lib/ash/query/function/ago.ex (hunk 3 above)
Comment (zachdaniel) [reply lib/ash/query/function/ago.ex:49]: I'd be fine with that.

**Comment by zachdaniel** [comment]
@mbuhot you ready for this to be merged?

**Comment by mbuhot** [comment]
@zachdaniel Yes I think it's good to merge now. Thanks.

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀
